File: 프로그래머스/5/49190. 방의 개수/방의 개수.py
def solution(arrows):
    op=[[-1,0],[-1,1],[0,1],[1,1],[1,0],[1,-1],[0,-1],[-1,-1]]
    n=0 
    e=0
    x,y=0,0
    node=dict()
    node[(0,0)]=0
    edge=dict()
    for i in arrows:
        nx=x+op[i][0]
        ny=y+op[i][1] 
        ni=(i+4)%8
        if (nx,ny,ni) not in edge and (x,y,i) not in edge:
            
            if i%2==1:   
                i1=(i-1)
                i2=(i+1)%8
                i3=(i1+3)%8
                i4=(i2+5)%8
                if (x+op[i1][0],y+op[i1][1],i3) in edge or (x+op[i2][0],y+op[i2][1],i4) in edge:
                    e+=2
                    n+=1

            node[(nx,ny)]=0
            edge[(x,y,i)]=0

        x+=op[i][0]
        y+=op[i][1]

    e+=len(edge)
    n+=len(node)
    return 1-n+e

File: 프로그래머스/5/49190. 방의 개수/방의 개수.cpp
#include <string>
#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <unordered_map>

using namespace std;

int dirr[] = { -1, -1, 0, 1, 1, 1, 0, -1 };
int dirc[] = { 0, 1, 1, 1, 0, -1, -1, -1 };

vector<pair<int, int>> vertex;
map<pair<int, int>, int> graph;
map<pair<int, int>, int> vis_node;
map<pair<int, int>, int> vis_edge;

int solution(vector<int> arrows) {
    int answer = 0;
    int sr = 0, sc = 0;

    vertex.push_back({sr, sc});
    for (int i = 0; i < arrows.size(); i++) {
        sr += dirr[arrows[i]];
        sc += dirc[arrows[i]];
        vertex.push_back({sr, sc});

        sr += dirr[arrows[i]];
        sc += dirc[arrows[i]];
        vertex.push_back({sr, sc});
    }

    int index = 0;
    for (auto& p : vertex) {
        if (graph.find(p) != graph.end()) {

        } else {
            graph[p] = index;
            index++;
        }
    }

    vis_node[vertex[0]] = 1;

    for (int i = 1; i < vertex.size(); i++) {
        pair<int, int> cp = vertex[i];
        pair<int, int> pp = vertex[i - 1];

        int cn = graph[cp];
        int pn = graph[pp];

        if (vis_node.find(cp) != vis_node.end() && vis_edge[{cn, pn}] == 0 && vis_edge[{pn, cn}] == 0) {
            answer++;
        }

        vis_edge[{cn, pn}]++;
        vis_edge[{pn, cn}]++;
        vis_node[cp] = 1;
    }

    return answer;
}